File: golang/dp_multidimensional/triangle.go
package main

func minimumTotal(triangle [][]int) int {
	if len(triangle) == 0 {
		return 0
	}

	dp := make([]int, len(triangle))

	copy(dp, triangle[len(triangle)-1])

	for i := len(triangle) - 2; i >= 0; i-- {
		for j := 0; j <= i; j++ {
			dp[j] = triangle[i][j] + min(dp[j], dp[j+1])
		}
	}

	return dp[0]
}

File: java/src/main/java/com/sanedge/interview/dp_multidimensional/Triangle.java
package com.sanedge.interview.dp_multidimensional;

import java.util.Arrays;

public class Triangle {

    public int minimumTotal(int[][] triangle) {
        if (triangle.length == 0) {
            return 0;
        }

        int[] dp = Arrays.copyOf(
            triangle[triangle.length - 1],
            triangle.length
        );

        for (int i = triangle.length - 2; i >= 0; i--) {
            for (int j = 0; j <= i; j++) {
                dp[j] = triangle[i][j] + Math.min(dp[j], dp[j + 1]);
            }
        }

        return dp[0];
    }
}
